addons/sale/migrations/11.0.1.1/post-migration.py

```diff
@@ -19,9 +20,107 @@ def activate_proforma(env):
+            GROUP BY rel.order_line_id,
+                ail.currency_id
+        ) AS sub
+        WHERE sol.id = sub.order_line_id
+            AND sol.currency_id = sub.currency_id
+        """,
+    )
+    # Now we set the amt_to_invoice with this simplified computation, but
+    # before discounting refunded invoiced. We apply a fixed rounding for
+    # avoiding resolution problems. It also uses a trick using the same table
+    # 2 times for avoiding to perform a subquery.
+    openupgrade.logged_query(
+        env.cr, """
+        UPDATE sale_order_line sol
+        SET amt_to_invoice = ROUND(
+            (CASE WHEN COALESCE(pt.invoice_policy, 'x') = 'delivery'
+             THEN sol.price_reduce_taxinc * sol.qty_delivered
+             ELSE sol.price_total END) - sol.amt_invoiced, 5
+        )
+        FROM sale_order_line sol2
+        LEFT JOIN product_product pp ON pp.id = sol2.product_id
+        LEFT JOIN product_template pt ON pt.id = pp.product_tmpl_id
+        WHERE sol2.id = sol.id
+            AND sol.state IN ('sale', 'done')
+        """,
+    )
+    # We have to do this second query that reduces the invoiced amount due to
+    # the refunded amount as the computation mixes several tables that can't
+    # be combined together. The computation is not exact, as we reduce amount
+    # for all lines with same product, but it's not even correct on upstream,
+    # as there's no real link.
+    # See https://github.com/odoo/odoo/pull/26054#pullrequestreview-143472733
+    openupgrade.logged_query(
+        env.cr, """
+        UPDATE sale_order_line sol
+        SET amt_invoiced = sol.amt_invoiced - sub.amount
+        FROM (
+            SELECT sol.id,
+                SUM(ail.price_total) AS amount
+            FROM sale_order_line sol
+                INNER JOIN sale_order_line_invoice_rel rel
+                    ON sol.id = rel.order_line_id
+                INNER JOIN account_invoice_line ail
+                    ON rel.invoice_line_id = ail.id
+                INNER JOIN account_invoice ai ON ai.id = ail.invoice_id
+                INNER JOIN account_invoice ai2 ON ai2.refund_invoice_id = ai.id
+            WHERE ai.state IN ('open', 'paid')
+                AND ai2.state IN ('open', 'paid')
+                AND ail.product_id = sol.product_id
+            GROUP BY sol.id
+        ) AS sub
+        WHERE sol.id = sub.id""",
+    )
+    # Finally, launch an ORM computation of lines with different invoice
+    # currency than sales order one
+    env.cr.execute(
+        """SELECT ail.id
+        FROM account_invoice_line ail
+        INNER JOIN sale_order_line_invoice_rel rel
+            ON rel.invoice_line_id = ail.id
+        INNER JOIN sale_order_line sol
+            on sol.id = rel.order_line_id
+        WHERE ail.currency_id != sol.currency_id
+        """
+    )
+    inv_lines = env['account.invoice.line'].browse([
+        x[0] for x in env.cr.fetchall()
+    ])
+    inv_lines._compute_invoice_amount()
```

Comment:
You should extract the sale order lines instead, not the invoice lines. This method is from the so lines.

Reply:
Thanks for fixing this. I'm thinking that we don't need to link `sale_order_line` table, as we already have the ID in sale_order_line_invoice_rel.order_line_id.

Would you?

Reply:
hahaha

Reply:
but you need the sale_order_line table for the currency!!!

Reply:
Ah, right, my first intention was correct then, as I haven't taken that into account this second sight